File: src/services/creative-project.service.ts
import { callTauri } from "./tauri";

export interface CreativeProject {
  id: string;
  title: string;
  description: string | null;
  status: string;
  settingsJson: string | null;
  budgetJson: string | null;
  createdAt: string;
  updatedAt: string;
  archivedAt: string | null;
}

export interface UpsertCreativeProjectInput {
  id: string;
  title: string;
  description?: string | null;
  status?: string | null;
  settingsJson?: string | null;
  budgetJson?: string | null;
  archivedAt?: string | null;
}

export interface ListCreativeProjectsFilter {
  status?: string | null;
  limit?: number | null;
  offset?: number | null;
}

export const creativeProjectService = {
  upsertProject: (input: UpsertCreativeProjectInput) =>
    callTauri<CreativeProject>("upsert_creative_project", { input }),
  getProject: (id: string) => callTauri<CreativeProject | null>("get_creative_project", { id }),
  listProjects: (filter: ListCreativeProjectsFilter = {}) =>
    callTauri<CreativeProject[]>("list_creative_projects", { filter }),
};

File: src-tauri/src/services/creative_project_service.rs
use crate::infra::creative_project_repo;
use crate::infra::creative_types::{
    CreativeProject, ListCreativeProjectsFilter, UpsertCreativeProjectInput,
};
use crate::infra::path::PathProvider;
use crate::infra::AppResult;

pub struct CreativeProjectService {
    path_provider: PathProvider,
}

impl CreativeProjectService {
    pub fn new(path_provider: PathProvider) -> Self {
        Self { path_provider }
    }

    pub fn upsert_project(&self, input: UpsertCreativeProjectInput) -> AppResult<CreativeProject> {
        creative_project_repo::upsert_project(&self.db_path()?, input)
    }

    pub fn get_project(&self, id: &str) -> AppResult<Option<CreativeProject>> {
        creative_project_repo::get_project(&self.db_path()?, id)
    }

    pub fn list_projects(
        &self,
        filter: ListCreativeProjectsFilter,
    ) -> AppResult<Vec<CreativeProject>> {
        creative_project_repo::list_projects(&self.db_path()?, filter)
    }

    fn db_path(&self) -> AppResult<std::path::PathBuf> {
        self.path_provider.get_db_file_path()
    }
}
